shm/include/locators.hpp

```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "time.hpp"
+
+class TimeServiceLocator {
+public:
+    static Time* getTime();
+    static void provide(Time* time);
+
+private:
+    inline static Time* time_ = nullptr;
```

Comment:
I would move initialization to a .cpp file, to keep with the convention that only declarations stay in the .hpp
Additionally I think it's quite good idea to change:
`Time* time_` to `std::unique_ptr<Time> time_`

If we set the time_ from the beginning like this:
`time_ = std::make_unique<Time>(Time());` 
we could skip the f-ction `provide(Time* time)` [I tested such solution and it works]

Reply:
How, It's static member & class - when you will be initializing it? We do not have constructor here. We cannot initialize time "just like that" because its private member. Anyway I do not understand how you want to instantiate static class. Service Locators was made according to http://gameprogrammingpatterns.com/service-locator.html.According to unique pointers - "implicit copy error has occured"

Reply:
Ahh ok I have checked your code - now i understand. Although it will be easier to test, we actually have no control on What "time" - or other observer collector is passed to game and other objects. We have "hardcoded it" by initializing static member. I will leave it as it is although testing is harder, We have some really basic IoC, because we can "inject" time. If we hide the "entry point" Our solution will be not configurable. For now i will leave it as it is, however I probably ask Mateusz what he thinks, because your solution has obvious advantages though...

Reply:
Hmmm, I think in our example we will not have any need to set another time or configure it. In my opinion we will always use this default Time object as our time. Perhaps only in a situation if we would like to restart the game - then we would need the possibility to set the time again in our TimeServiceLocator. However I think that we could mix both our solutions - set the time by default as I did and also keep the function provide time if we will ever need to change the time :)

Reply:
👍

Reply:
But we need to use it in few places. Even if we cover it In service locator, moving semantics will be horrible (please look that "game" needs time too - moreover - game provides time and inject it to ServiceLocator. How we can make it with hidden initialize in .cpp? Get? ok but right now unique pointer will have 2 owners and it will crash.

Reply:
Done, I hope you will be satisfied :)